radixdlt-java/src/main/java/com/radixdlt/client/core/fungible/FungibleParticleTransitioner.java

```diff
@@ -0,0 +1,146 @@
+		Function<List<U>, List<U>> transitionedCombiner,
+		BiFunction<UInt256, T, T> migrator,
+		Function<List<T>, List<T>> migratedCombiner,
+		Function<T, UInt256> amountMapper
+	) {
+		this.transitioner = transitioner;
+		this.transitionedCombiner = transitionedCombiner;
+		this.migrator = migrator;
+		this.migratedCombiner = migratedCombiner;
+		this.amountMapper = amountMapper;
+	}
+
+	public static class FungibleParticleTransition<T, U> {
+		private final ImmutableList<T> removed;
+		private final ImmutableList<T> migrated;
+		private final ImmutableList<U> transitioned;
+
+		public FungibleParticleTransition(
+			ImmutableList<T> removed,
+			ImmutableList<T> migrated,
+			ImmutableList<U> transitioned
+		) {
+			this.removed = removed;
+			this.migrated = migrated;
+			this.transitioned = transitioned;
+		}
+
+		public List<T> getRemoved() {
+			return removed;
+		}
+
+		public List<T> getMigrated() {
+			return migrated;
+		}
+
+		public List<U> getTransitioned() {
+			return transitioned;
+		}
+	}
+
+	private static class FungibleParticleTransitionBuilder<T, U> {
+		private ImmutableList.Builder<T> removedBuilder = ImmutableList.builder();
+		private List<T> migrated = new ArrayList<>();
+		private List<U> transitioned = new ArrayList<>();
+		private final Function<List<U>, List<U>> transitionedCombiner;
+		private final Function<List<T>, List<T>> migratedCombiner;
+
+		private FungibleParticleTransitionBuilder(
+			Function<List<U>, List<U>> transitionedCombiner,
+			Function<List<T>, List<T>> migratedCombiner
+		) {
+			this.transitionedCombiner = transitionedCombiner;
+			this.migratedCombiner = migratedCombiner;
+		}
+
+
+		private FungibleParticleTransitionBuilder<T, U> addTransitioned(U u) {
+			transitioned.add(u);
+			return this;
+		}
+
+		private FungibleParticleTransitionBuilder<T, U> addMigrated(T t) {
+			migrated.add(t);
+			return this;
+		}
+
+		private FungibleParticleTransitionBuilder<T, U> addRemoved(T t) {
+			removedBuilder.add(t);
+			return this;
+		}
+
+		private FungibleParticleTransition<T, U> build() {
+			return new FungibleParticleTransition<>(
+				removedBuilder.build(),
+				ImmutableList.copyOf(migratedCombiner.apply(migrated)),
+				ImmutableList.copyOf(transitionedCombiner.apply(transitioned)));
+		}
+	}
+
+	public FungibleParticleTransition<T, U> createTransition(
+		List<T> unconsumedFungibles,
+		UInt256 toAmount
+	) {
+		UInt256 balance = unconsumedFungibles.stream().map(amountMapper).reduce(UInt256.ZERO, UInt256::add);
+		if (balance.compareTo(toAmount) < 0) {
+			throw new RuntimeException("Not enough to cannot create transition");
+		}
+
+		UInt256 consumerTotal = UInt256.ZERO;
+		Iterator<T> iterator = unconsumedFungibles.iterator();
+
+		final FungibleParticleTransitionBuilder<T, U> transitionBuilder = new FungibleParticleTransitionBuilder<>(
+			transitionedCombiner,
+			migratedCombiner
+		);
+
+		while (consumerTotal.compareTo(toAmount) < 0 && iterator.hasNext()) {
```

Comment:
I assume this is all deterministic, right? I can't find any source of non-determinism in the transition logic, just want to double-check, as that might cause some head scratches if not.

Reply:
yes deterministic